lib/wasm/cache.js

```diff
@@ -0,0 +1,152 @@
+// Copyright Joyent, Inc. and other Node contributors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to permit
+// persons to whom the Software is furnished to do so, subject to the
+// following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+'use strict';
+
+const crypto = require('crypto');
+const fs = require('fs');
+const fsp = require('fs/promises');
+const os = require('os');
+const path = require('path');
+
+const { Promise, PromiseAll } = primordials;
+
+let cacheSubdir = '.cache/node/wasm-cache';
+switch (os.platform()) {
+  case 'darwin':
+    cacheSubdir = 'Library/Caches/node/wasm-cache';
+    break;
+  case 'win32':
+    cacheSubdir = 'AppData\\Local\\node\\wasm-cache';
+    break;
+}
+
+let cacheDir = path.resolve(os.homedir(), cacheSubdir);
```

Comment:
I see where @devsnek’s concerns are coming from – this is mixing some pretty specific I/O details with low-level functionality.

Ideally, we’d expose an API that doesn’t do any I/O, and lets the user fill that part in by specifying their own callback(s?) for that – how would that sound to you?

Reply:
@addaleax Thank you for taking look! I have a meta question: what is the best channel to discuss the feature scope / public facing APIs?

Reply:
I’m not sure, but probably this PR itself – although I can see that this might turn into a long discussion…

Reply:
@addaleax 
> Ideally, we’d expose an API that doesn’t do any I/O, and lets the user fill that part in by specifying their own callback(s?)

We could do it differently:
  1. either provide a low-level WebAssembly `deserialise` primitive, let the user build on top,
  2. or build a higher-level api with the caching largely taken care of,
  3. something in-between?

Personally, I believe *2* is more valuable. Pros:
  * implementing caching properly is non-trivial (have to detect if the source file was changed and handle race conditions while updating the cache to name a few challenges);
  * cache files end up in predictable locations, it's easy to reclaim disk space and troubleshoot issues;
  * doesn't expose implementation details that are likely to change. 

Furthermore, sometimes the user can't be expected to write custom caching code — consider `require`-ing a WASI module or Emscripten-generated wrapper.

@kripken

Reply:
I think your first two Pros are really Pros for option 1? Implementing caching properly is non-trivial – hence, it’s best if a userland package addresses this, instead of adding the complexity to Node.js core and tying updates to our release schedule. And while cache files may end up in predictable locations, what those locations are may depend strongly on different user’s needs – also a good point for moving this into userland modules.

Reply:
> it’s best if a userland package addresses this, instead of adding the complexity to Node.js core and tying updates to our release schedule 


These are valid points. However It looks like the core itself could benefit from the cache due to this feature
https://nodejs.org/api/all.html#esm_wasm_modules

FYI, concerning the cache there’s some prior art in Python we could borrow from.

https://www.python.org/dev/peps/pep-3147/

https://docs.python.org/3/whatsnew/3.8.html#parallel-filesystem-cache-for-compiled-bytecode-files


Personally, I’m mostly interested in Emscripten use case. Having a minimal low-level api in core will work for me. So if you feel strongly about it let’s do it this way.

Reply:
@addaleax @devsnek 

What if we expose WasmStreaming to userland without inventing new interfaces? 
```js
class Foo { ... }
require('v8').registerWebAssemblyStreamer(Foo, (foo, wasmStreaming) => {
  wasmStreaming.setCompiledModuleBytes(...);
  wasmStreaming.setUrl(...);
  wasmStreaming.onBytesReceived(...);
  wasmStreaming.finish();
});
WebAssembly.compileStreaming(new Foo(...));
```